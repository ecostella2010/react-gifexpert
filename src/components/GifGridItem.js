import React from 'react'
import PropType from 'prop-types';

export const GifGridItem = ( { title, url } ) => {

    //console.log ({id, title, url});

    return (
        <div className = "card animate__animated animate__bounce animate__fadeIn">
            <img src={ url } alt = { title }></img>
            <p> { title } </p>
            {/* { title } */}
        </div>
    )
}

GifGridItem.propType = {
    title: PropType.string.isRequired,
    url: PropType.string.isRequired
                 
}