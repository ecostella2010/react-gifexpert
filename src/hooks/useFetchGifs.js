import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";


export const useFetchGifs = ( category ) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect ( () => {
        //getGifs ()
        //getGifs( category )
        // .then(imgs => setImages( imgs ))
        //.then( setImages )
        //}, [ category ]);

        getGifs ( category ) 
        .then( imgs => { setState   ({
                                        data: imgs,
                                        loading: false
                                    });
                        });    
    }, [ category ]);

    // setTimeout(() => {
    //     setState({
    //         data: [1,2,3,4],
    //         loading: false
    //     });
    // }, 3000);

    return state;  // { data:[], loading: true };
}