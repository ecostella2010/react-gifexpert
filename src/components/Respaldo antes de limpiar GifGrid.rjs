//import React, { useState, useEffect } from 'react'
import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ( { category } ) => {

    //const [count, setCount] = useState(0);
    //Casi como el component dispose para inicializar

    // const [images, setImages] = useState([]);

    // const state = useFetchGifs();
    // console.log( state );
    //Al desestructurar
    //const { data, loading } = useFetchGifs( category );

    //Si queremos desestructurar renombrando
    const { data:images, loading } = useFetchGifs( category );

    // console.log( loading );
    // console.log( data );

    //Se lleva a custome hook useFetchGifs
    // useEffect ( () => {
    //     //getGifs ()
    //     getGifs( category )
    //     // .then(imgs => setImages( imgs ))
    //     .then( setImages )
    // }, [ category ]);

    // const getGifs = async() => {
    //     //console.log('category:', category);
    //     const url = `https://api.giphy.com/v1/gifs/search?q=${ encodeURI(category) }&limit=10&api_key=HQYcz1cE4MrSnWH5veByJYYTwqBXu7zz`;
    //     //console.log (url);
    //     const resp = await fetch ( url );
    //     //const data = await resp.json();
    //     //Uso RestructuraciÃ³n 
    //     const { data } = await resp.json();
    //     //console.log( data );
    //     const gifs = data.map ( img => {
    //         return {
    //             id : img.id,
    //             title : img.title,
    //             url : img.images?.downsized_medium.url
    //         }
    //     });

    //     //console.log( gifs );
    //     setImages(gifs);
    //     //console.log( images );
        
    // }

    return (
        <>
            <h3> Filtro : { category } </h3>

            {/* { loading ? 'Cargando...' : 'Data cargada' } */}

            {/* { loading ? <p> Loading... </p> : null } */}
            { loading && <p> Loading... </p> }

            <div className="card-grid">
                {
                    images.map ( img  => (
                            <GifGridItem
                                key = { img.id } 
                                // img = { img } 
                                //Para enviar todas las propiedades como propiedades independientes 
                                {...img}
                                
                            />
                    ))
                }

            </div>
        </>

        // <div>
        //     <h3> { category } </h3>
        //     <ol>
        //         {
        //             // images.map ( img => (
        //             //      <li key = { img.id } > { img.title }  </li>    
        //             // ) )

        //             //Desestructurando
        //             images.map ( ({ id, title }) => (
        //                 <li key = { id } > { title }  </li>    
        //            ) )

        //         }
        //     </ol>

        //     {/* <h3> { count } </h3>
        //     <button onClick= { () => setCount( count + 1) } > </button> */}
        // </div>
    )
}