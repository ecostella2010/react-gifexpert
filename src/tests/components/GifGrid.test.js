import React from 'react'; 
import { shallow } from 'enzyme';

import '@testing-library/jest-dom';
import { GifGrid } from '../../components/gifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Prueba en GifGrid', () => {
    const category = 'One Punch';

    //let wrapper = shallow( <GifGrid category = { category } />);

    // beforeEach ( () => {
    //     jest.clearAllMocks();
    //     wrapper = shallow( <GifGrid category = { category } />);
    // });
    
    test('debe de mostar < GifGrid /> correstamente', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <GifGrid category = { category } />);

        expect( wrapper ).toMatchSnapshot();
    });


    test('debe de mostrar items cuando se cargan imagenes suseFetchGifs', () => {

        const  gifs = [
            {
                id: 'ABC',
                url: 'https://localhost/cualquie/cosa.jpg',
                title: 'Cualquier cosa'
            },
            {
                id: '123',
                url: 'https://localhost/cualquie/cosa.jpg',
                title: 'Cualquier cosa'
            }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow( <GifGrid category = { category } />);
        //expect( wrapper ).toMatchSnapshot();
        expect ( wrapper.find('p').exists() ).toBe( false );
        expect ( wrapper.find('GifGridItem').length ).toBe( gifs.length );

       
    })
    

});
