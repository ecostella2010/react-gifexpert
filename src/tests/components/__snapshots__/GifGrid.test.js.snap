// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en GifGrid debe de mostar < GifGrid /> correstamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounce animate__fadeIn"
  >
     Filtro : 
    One Punch
     
  </h3>
  <p
    className="animate__animated animate__bounce animate__flash"
  >
     Loading... 
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
